### PR TITLE
feat(next): add payment element and submit

### PR DESCRIPTION
**Pull request by StaberindeZA**
## Because

* Add the Stripe Payment Element to the checkout page
* Be able to checkout and successfully add a new payment method and new subscription.

## This pull request

* Adds the PaymentElement to the StripeWrapper
* Adds the Subscribe Now button
* Adds the Full Name field
* Updates the NestApp with Firestore and StatsD provider factories

## Issue that this pull request solves

Closes: #FXA-8850 and #FXA-7810

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [x] I have added necessary documentation (if appropriate).
- [x] I have verified that my changes render correctly in RTL (if appropriate).

## Additional notes

This PR does not currently result in a successful subscription creation. There's still some outstanding work to be completed in M3a and M3b before that can be tested.

### CODE REVIEW

File: apps/payments/next/middleware.ts
Comment (pdehaan) [inline apps/payments/next/middleware.ts:13]: nit: should this be production AND stage?

File: libs/payments/ui/src/lib/client/components/CheckoutForm.tsx
Comment (pdehaan) [inline libs/payments/ui/src/lib/client/components/CheckoutForm.tsx:118]: There is a `z-50` on https://tailwindcss.com/docs/z-index. 

#### Checking for custom `z-[nnn]` values
<details>
<summary><kbd>git grep -IPn "\bz-\[\d+\]" | cat</kbd></summary>

```sh
apps/payments/next/app/styles/global.css:62:  @apply payment-button-disabled cursor-not-allowed after:bg-white after:opacity-50 after:z-[5];
libs/payments/ui/src/lib/client/components/CheckoutForm.tsx:118:        className="flex items-center justify-center bg-blue-500 font-semibold h-12 rounded-md text-white w-full p-4 mt-6 disabled:relative disabled:after:absolute disabled:after:content-[''] disabled:after:top-0 disabled:after:left-0 disabled:after:w-full disabled:after:h-full disabled:after:bg-white disabled:after:opacity-50 disabled:after:z-[100] disabled:border-none"
packages/fxa-payments-server/src/components/LoadingOverlay/index.tsx:14:      className="flex justify-center items-center fixed inset-0 z-[9999]"
packages/fxa-payments-server/src/components/PayPalButton/index.tsx:250:    ' payment-button-disabled after:bg-white after:opacity-40 after:z-[100]';
packages/fxa-payments-server/src/components/PaymentForm/index.tsx:271:    'payment-button-disabled after:bg-white after:opacity-50 after:z-[100] border-none';
packages/fxa-payments-server/src/components/fields/index.tsx:452:        ' payment-button-disabled cursor-not-allowed after:bg-white after:opacity-50 z-[5] border-none focus:border-solid focus:border-2 focus:border-offset-2 focus:border-blue-400'
packages/fxa-payments-server/src/styles/tailwind.css:154:  @apply payment-button-disabled cursor-not-allowed after:bg-white after:opacity-50 after:z-[5];
packages/fxa-payments-server/src/styles/tailwind.css:160:  @apply z-[6] px-4 py-1.5 left-0 -mt-12;
packages/fxa-react/components/Portal/index.tsx:44:        'portal fixed top-0 inset-x-0 p-0 z-[9999] tablet:px-16'
packages/fxa-settings/src/components/Settings/HeaderLockup/index.tsx:28:        className="desktop:hidden me-6 p-2 self-center -m-2 z-[1] rounded hover:bg-grey-100"
packages/fxa-settings/src/components/Settings/HeaderLockup/index.tsx:70:        className="inline-block relative p-2 -m-2 z-[1] rounded hover:bg-grey-100"
```
</details>

---

#### Checking for Tailwind `z-nnn` values
<details>
<summary><kbd>git grep -IPn "\bz-\d+" -- ":(exclude)*.txt" | cat</kbd></summary>

```sh
apps/payments/next/app/layout.tsx:25:          className="bg-white fixed flex justify-between items-center shadow h-16 left-0 top-0 mx-auto my-0 px-4 py-0 w-full z-10 tablet:h-20"
libs/shared/assets/src/styles/tooltips.css:6:  @apply absolute text-start w-[calc(100%+.5rem)] -left-1 desktop:w-64 desktop:left-auto px-5 py-4 bg-white rounded-sm top-[calc(100%+.25rem)] desktop:-top-2 border border-grey-100 shadow-sm transition-opacity z-10;
packages/fxa-admin-panel/src/components/PageAccountSearch/index.tsx:105:        className="p-2 border-b border-grey-100 block hover:bg-grey-10 focus:bg-grey-10 z-50"
packages/fxa-admin-panel/src/components/PageAccountSearch/index.tsx:172:              className="suggestions-list absolute top-full w-full bg-white border border-grey-100 mt-3 shadow-sm rounded overflow-hidden z-50"
packages/fxa-payments-server/src/components/Header/index.tsx:30:      className={`${className} flex justify-between items-center bg-white shadow h-16 fixed left-0 top-0 my-0 mx-auto px-4 py-0 w-full z-10 tablet:h-20`}
packages/fxa-payments-server/src/components/PayPalButton/index.tsx:259:          className={`paypal-button w-60 m-auto mt-14 mb-16 relative z-0${
packages/fxa-react/components/Header/index.tsx:19:      className={props.className || 'sticky top-0 bg-grey-10 z-10'}
packages/fxa-react/styles/tooltips.css:6:  @apply absolute text-start w-[calc(100%+.5rem)] -left-1 desktop:w-64 desktop:left-auto px-5 py-4 bg-white rounded-sm top-[calc(100%+.25rem)] desktop:-top-2 border border-grey-100 shadow-sm transition-opacity z-10;
packages/fxa-settings/.storybook/design-guide/Copiable.tsx:19:      <button className="copy-content hidden absolute top-1/2 left-1/2 transform -translate-x-1/2 -translate-y-1/2 z-10 text-sm pointer bg-white text-grey-900 focus:bg-grey-100 px-1 shadow-md rounded border border-grey-200 rounded-md whitespace-nowrap font-mono pr-7">
packages/fxa-settings/.storybook/design-guide/Header.tsx:6:    <div className="bg-grey-50 flex items-center justify-between sticky z-20 top-0 py-4 pb-3 border-b border-grey-400/25 px-2 -mt-4 -ml-12 -mr-12">
packages/fxa-settings/src/components/Settings/AlertBar/index.tsx:46:      className="fixed z-10"
packages/fxa-settings/src/components/Settings/BentoMenu/index.tsx:54:          className={`w-full h-full fixed top-0 start-0 bg-white z-10
packages/fxa-settings/src/components/Settings/ContentSkip/index.tsx:11:    className="sr-only focus:w-auto focus:h-auto focus:clip-auto absolute bg-white font-medium shadow-md px-3 py-2 z-10 rounded ml-2 mt-2 cursor-pointer"
packages/fxa-settings/src/components/Settings/Modal/index.tsx:61:        className="flex flex-col justify-center fixed inset-0 z-50 w-full px-2 h-full bg-black/75"
packages/fxa-settings/src/components/Tooltip/index.tsx:52:        `z-50 absolute py-2 px-6 text-center text-white
```
</details>

File: apps/payments/next/middleware.ts
Comment (StaberindeZA) [reply apps/payments/next/middleware.ts:13]: `stage` isn't one of the [standard NODE_ENV values recommended by Next.js](https://nextjs.org/docs/messages/non-standard-node-env), so we should be fine with just `production`.

File: libs/payments/ui/src/lib/client/components/CheckoutForm.tsx
Comment (StaberindeZA) [reply libs/payments/ui/src/lib/client/components/CheckoutForm.tsx]: I'll replace with z-20.

File: libs/payments/ui/src/lib/actions/checkoutCartWithStripe.ts
Comment (julianpoy) [inline libs/payments/ui/src/lib/actions/checkoutCartWithStripe.ts:25]: Imo this should be outside of the action itself. Action should only have the responsibility of dispatching the call to the server.

File: libs/shared/db/firestore/src/lib/firestore.provider.ts
Comment (julianpoy) [inline libs/shared/db/firestore/src/lib/firestore.provider.ts:72]: Can we call this `FirestoreProvider`?

File: libs/shared/metrics/statsd/src/lib/statsd.provider.ts
Comment (julianpoy) [inline libs/shared/metrics/statsd/src/lib/statsd.provider.ts:35]: Can we name this `StatsDProvider`, and rename `MockStatsDFactory` to `MockStatsDProvider`?

**Review by julianpoy — APPROVED** [review]

File: libs/payments/ui/src/lib/client/components/StripeWrapper.tsx
Comment (pdehaan) [inline libs/payments/ui/src/lib/client/components/StripeWrapper.tsx:50]: ~~Per https://tailwindcss.com/docs/customizing-colors#default-color-palette, it looks like `gray-400` is `#9ca3af`.~~

Nevermind, you said we're overriding the default colors in https://github.com/mozilla/fxa/blob/ce46102c869a41ed7d0e14320c95084da2dc54de/libs/shared/assets/src/configs/tailwind.config.js#L168

File: libs/shared/contentful/src/lib/contentful.client.ts
Comment (pdehaan) [inline libs/shared/contentful/src/lib/contentful.client.ts:1]: nit: Do we have a Jira ticket number for this yet?

File: libs/shared/metrics/statsd/src/lib/statsd.config.ts
Comment (pdehaan) [inline libs/shared/metrics/statsd/src/lib/statsd.config.ts:19]: nit: should this be `@IsInt()`?

File: libs/payments/ui/src/lib/client/components/CheckoutForm.tsx
Comment (StaberindeZA) [reply libs/payments/ui/src/lib/client/components/CheckoutForm.tsx]: Updated this to more closely match the Stripe components which resulted in the following. 
* removed `h-x`
* removed `focus:border-[#0a84ff]`
* Updated placeholder text to use `grey-500`

File: libs/payments/ui/src/lib/client/components/CheckoutForm.tsx
Comment (StaberindeZA) [reply libs/payments/ui/src/lib/client/components/CheckoutForm.tsx:138]: They should, but localization for client components is currently blocked by #16595 , and will be added as a polish to this PR once the blocker lands.

File: libs/shared/contentful/src/lib/contentful.client.ts
Comment (StaberindeZA) [reply libs/shared/contentful/src/lib/contentful.client.ts:1]: Not yet, in other comments I mentioned this will be picked up in the next Milestone, which hasn't started planning yet.

File: libs/payments/ui/src/lib/client/components/CheckoutForm.tsx
Comment (xlisachan) [inline libs/payments/ui/src/lib/client/components/CheckoutForm.tsx:74]: :eyes:

File: libs/payments/ui/src/lib/client/components/CheckoutForm.tsx
Comment (xlisachan) [inline libs/payments/ui/src/lib/client/components/CheckoutForm.tsx:131]: :eyes:

File: libs/payments/ui/src/lib/client/components/CheckoutForm.tsx
Comment (StaberindeZA) [reply libs/payments/ui/src/lib/client/components/CheckoutForm.tsx:166]: Funnily enough I just had this conversation with @xlisachan and discussed how to handle this. 

I'd like to go with Tailwind recommendation, probably they know something we don't, and we discussed pulling this out into a separate component once we add more buttons to the page.

I created [FXA-9625](https://mozilla-hub.atlassian.net/browse/FXA-9625) for some general cleanup and included that work in this ticket as well.

**Review by xlisachan — APPROVED** [review]
r+ thanks Reino!

**Pull request merged**
